File: app/src/main/java/com/gasparaiciukas/owntrainer/activity/SettingsActivity.java
package com.gasparaiciukas.owntrainer.activity;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.AutoCompleteTextView;

import com.gasparaiciukas.owntrainer.R;
import com.gasparaiciukas.owntrainer.database.User;
import com.google.android.material.appbar.MaterialToolbar;
import com.google.android.material.textfield.TextInputEditText;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import io.realm.Realm;

public class SettingsActivity extends AppCompatActivity {

    // Variables
    private String sex;
    private int age;
    private int height;
    private double weight;
    private String lifestyle;

    // Text views
    private AutoCompleteTextView sexMenu;
    private TextInputEditText tAge;
    private TextInputEditText tHeight;
    private TextInputEditText tWeight;
    private AutoCompleteTextView lifestyleMenu;

    // Database
    private Realm realm;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_settings);
        initUi();
        loadData();
    }

    private void initUi() {
        // Back button
        MaterialToolbar toolbar = findViewById(R.id.settings_top_app_bar);
        toolbar.setNavigationOnClickListener(v -> onBackPressed());

        // Options
        sexMenu = findViewById(R.id.settings_sex_menu);
        tAge = findViewById(R.id.settings_age_text);
        tHeight = findViewById(R.id.settings_height_text);
        tWeight = findViewById(R.id.settings_weight_text);
        lifestyleMenu = findViewById(R.id.settings_lifestyle_menu);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        writeUserToDatabase();
    }

    private void writeUserToDatabase() {
        // Write to database
        realm = Realm.getDefaultInstance();
        User user = new User();
        user.setUserId("user");
        user.setAgeInYears(age);
        user.setSex(sex);
        user.setHeightInCm(height);
        user.setWeightInKg(weight);
        user.setLifestyle(lifestyle);
        user.setStepLengthInCm(user.calculateStepLengthInCm(height, sex));
        user.setBmr(user.calculateBmr(weight, height, age, sex));
        user.setKcalBurnedPerStep(user.calculateKcalBurnedPerStep(weight, height, user.getAvgWalkingSpeedInKmH()));
        user.setDailyKcalIntake(user.calculateDailyKcalIntake(user.getBmr(), lifestyle));
        user.setDailyCarbsIntakeInG(user.calculateDailyCarbsIntake(user.getDailyKcalIntake()));
        user.setDailyFatIntakeInG(user.calculateDailyFatIntake(user.getDailyKcalIntake()));
        user.setDailyProteinIntakeInG(user.calculateDailyProteinIntakeInG(weight));

        realm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(@NotNull Realm realm) {
                realm.insertOrUpdate(user);
            }
        });
        realm.close();
    }

    private void loadData() {
        realm = Realm.getDefaultInstance();
        User u = realm.where(User.class)
                .equalTo("userId", "user")
                .findFirst();
        sex = u.getSex();
        age = u.getAgeInYears();
        height = u.getHeightInCm();
        weight = u.getWeightInKg();
        lifestyle = u.getLifestyle();
        realm.close();

        // Insert current data into fields
        sexMenu.setText(sex);
        tAge.setText(String.valueOf(age));
        tHeight.setText(String.valueOf(height));
        tWeight.setText(String.valueOf(weight));
        lifestyleMenu.setText(lifestyle);

        // Set up listeners
        List<String> sexList = new ArrayList<>(Arrays.asList("Male", "Female"));
        ArrayAdapter sexAdapter = new ArrayAdapter(this, R.layout.details_list_item, sexList);
        sexMenu.setAdapter(sexAdapter);
        sexMenu.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                sex = sexMenu.getText().toString();
            }
        });
        tAge.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                // do nothing
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                // do nothing
            }

            @Override
            public void afterTextChanged(Editable s) {
                if (!s.toString().isEmpty())
                    age = Integer.parseInt(s.toString());
            }
        });
        tHeight.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                // do nothing
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                // do nothing
            }

            @Override
            public void afterTextChanged(Editable s) {
                if (!s.toString().isEmpty())
                    height = Integer.parseInt(s.toString());
            }
        });
        tWeight.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                // do nothing
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                // do nothing
            }

            @Override
            public void afterTextChanged(Editable s) {
                if (!s.toString().isEmpty())
                    weight = Double.parseDouble(s.toString());
            }
        });
        lifestyleMenu.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                lifestyle = lifestyleMenu.getText().toString();
            }
        });

        List<String> lifestyleList = new ArrayList<>(Arrays.asList(
                "Sedentary",
                "Lightly active",
                "Moderately active",
                "Very active",
                "Extra active"));
        ArrayAdapter lifestyleAdapter = new ArrayAdapter(this, R.layout.details_list_item, lifestyleList);
        lifestyleMenu.setAdapter(lifestyleAdapter);
    }

}

File: app/src/main/java/com/gasparaiciukas/owntrainer/activity/SettingsActivity.kt
package com.gasparaiciukas.owntrainer.activity

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.widget.AdapterView.OnItemClickListener
import android.widget.ArrayAdapter
import androidx.appcompat.app.AppCompatActivity
import com.gasparaiciukas.owntrainer.R
import com.gasparaiciukas.owntrainer.database.User
import com.gasparaiciukas.owntrainer.databinding.ActivitySettingsBinding
import io.realm.Realm
import java.util.*

class SettingsActivity : AppCompatActivity() {
    private lateinit var binding: ActivitySettingsBinding

    // Variables
    private lateinit var sex: String
    private var age = 0
    private var height = 0
    private var weight = 0.0
    private lateinit var lifestyle: String

    // Database
    private lateinit var realm: Realm

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivitySettingsBinding.inflate(layoutInflater)
        setContentView(binding.root)
        binding.appBar.setNavigationOnClickListener { onBackPressed() }
        loadData()
    }

    override fun onDestroy() {
        super.onDestroy()
        writeUserToDatabase()
    }

    private fun writeUserToDatabase() {
        // Write to database
        realm = Realm.getDefaultInstance()
        val user = User()
        user.userId = "user"
        user.ageInYears = age
        user.sex = sex
        user.heightInCm = height
        user.weightInKg = weight
        user.lifestyle = lifestyle
        user.stepLengthInCm = user.calculateStepLengthInCm(height.toDouble(), sex)
        user.bmr = user.calculateBmr(weight, height.toDouble(), age, sex)
        user.kcalBurnedPerStep =
            user.calculateKcalBurnedPerStep(weight, height.toDouble(), user.avgWalkingSpeedInKmH)
        user.dailyKcalIntake = user.calculateDailyKcalIntake(user.bmr, lifestyle)
        user.dailyCarbsIntakeInG = user.calculateDailyCarbsIntake(user.dailyKcalIntake)
        user.dailyFatIntakeInG = user.calculateDailyFatIntake(user.dailyKcalIntake)
        user.dailyProteinIntakeInG = user.calculateDailyProteinIntakeInG(weight)
        realm.executeTransaction { r -> r.insertOrUpdate(user) }
        realm.close()
    }

    private fun loadData() {
        realm = Realm.getDefaultInstance()
        val user = realm.where(User::class.java)
            .equalTo("userId", "user")
            .findFirst()
        if (user != null) {
            sex = user.sex
            age = user.ageInYears
            height = user.heightInCm
            weight = user.weightInKg
            lifestyle = user.lifestyle
        }
        realm.close()

        // Insert current data into fields
        binding.etSex.setText(sex)
        binding.etAge.setText(age.toString())
        binding.etHeight.setText(height.toString())
        binding.etWeight.setText(weight.toString())
        binding.etLifestyle.setText(lifestyle)

        // Set up listeners
        val sexList: List<String?> = ArrayList(listOf("Male", "Female"))
        val sexAdapter: ArrayAdapter<*> =
            ArrayAdapter<Any?>(this, R.layout.details_list_item, sexList)
        binding.etSex.setAdapter(sexAdapter)
        binding.etSex.onItemClickListener =
            OnItemClickListener { _, _, _, _ -> sex = binding.etSex.text.toString() }
        binding.etAge.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
                // do nothing
            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                // do nothing
            }

            override fun afterTextChanged(s: Editable) {
                if (s.toString().isNotEmpty()) age = s.toString().toInt()
            }
        })
        binding.etHeight.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
                // do nothing
            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                // do nothing
            }

            override fun afterTextChanged(s: Editable) {
                if (s.toString().isNotEmpty()) height = s.toString().toInt()
            }
        })
        binding.etWeight.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
                // do nothing
            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                // do nothing
            }

            override fun afterTextChanged(s: Editable) {
                if (s.toString().isNotEmpty()) weight = s.toString().toDouble()
            }
        })
        binding.etLifestyle.onItemClickListener =
            OnItemClickListener { _, _, _, _ -> lifestyle = binding.etLifestyle.text.toString()
        }
        val lifestyleList: List<String?> = ArrayList(
            listOf(
                "Sedentary",
                "Lightly active",
                "Moderately active",
                "Very active",
                "Extra active"
            )
        )
        val lifestyleAdapter: ArrayAdapter<*> =
            ArrayAdapter<Any?>(this, R.layout.details_list_item, lifestyleList)
        binding.etLifestyle.setAdapter(lifestyleAdapter)
    }
}